# 1. 导入一些必要的库

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import RobustScaler,LabelEncoder,MinMaxScaler,StandardScaler
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

# 2. 数据集导入

In [2]:
train = pd.read_csv('/home/mw/input/minsu2472/train_data.csv')
test = pd.read_csv('/home/mw/input/minsu2472/test_data.csv')

train = train.drop('邮编', axis=1)
test = test.drop('邮编', axis=1)

# 3. 数据探索

## 3.1 基本信息

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59288 entries, 0 to 59287
Data columns (total 26 columns):
Unnamed: 0     59288 non-null int64
数据ID           59288 non-null object
容纳人数           59288 non-null int64
便利设施           59288 non-null object
洗手间数量          59128 non-null float64
床的数量           59184 non-null float64
床的类型           59288 non-null int64
卧室数量           59210 non-null float64
取消条款           59288 non-null int64
所在城市           59288 non-null int64
清洁费            59288 non-null int64
首次评论日期         46586 non-null object
房主是否有个人资料图片    59128 non-null object
房主身份是否验证       59128 non-null object
房主回复率          44647 non-null object
何时成为房主         59128 non-null object
是否支持随即预订       59288 non-null int64
最近评论日期         46616 non-null object
维度             59288 non-null float64
经度             59288 non-null float64
民宿周边           53779 non-null object
评论个数           59288 non-null int64
房产类型           59288 non-null int64
民宿评分           45893 non-null float64
房型    

In [4]:
train.head()

Unnamed: 0     数据ID  容纳人数  \
0           0  train_0     4   
1           1  train_1     2   
2           2  train_2     4   
3           3  train_3     2   
4           4  train_4     3   

                                                便利设施  洗手间数量  床的数量  床的类型  卧室数量  \
0  {TV,"Cable TV",Internet,"Wireless Internet","A...    1.5   3.0     4   2.0   
1  {TV,"Wireless Internet",Kitchen,"Free parking ...    1.0   1.0     4   1.0   
2  {TV,"Air conditioning",Kitchen,Heating,"Smoke ...    1.0   2.0     4   0.0   
3                                                 {}    1.0   1.0     4   1.0   
4  {Internet,"Wireless Internet","Air conditionin...    1.0   1.0     4   1.0   

   取消条款  所在城市  清洁费      首次评论日期 房主是否有个人资料图片 房主身份是否验证 房主回复率      何时成为房主  \
0     0     3    0  2015-05-07           t        t   NaN  2015-02-25   
1     2     4    1  2016-07-02           t        t   NaN  2009-10-27   
2     2     4    1  2017-07-01           t        f  100%  2017-06-29   
3     0     5    1         NaN           t        t   NaN  2013-03-19   
4     1     4    1  2014-04-30           t        t  100%  2011-07-30   

   是否支持随即预订      最近评论日期         维度          经度                民宿周边  评论个数  \
0         0  2016-06-26  34.109039 -118.273390           Los Feliz    12   
1         1  2016-07-31  40.812897  -73.919163          Mott Haven     6   
2         1  2017-07-31  40.737643  -73.953309          Greenpoint     4   
3         0         NaN  37.759935 -122.420558    Mission District     0   
4         0  2016-05-22  40.683363  -73.949490  Bedford-Stuyvesant    16   

   房产类型  民宿评分  房型         价格  
0    17  97.0   0  64.918531  
1     0  87.0   0  54.918531  
2     0  80.0   0  73.219281  
3     0   NaN   1  64.093909  
4     0  99.0   0  68.454901

In [5]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14823 entries, 0 to 14822
Data columns (total 25 columns):
Unnamed: 0     14823 non-null int64
数据ID           14823 non-null object
容纳人数           14823 non-null int64
便利设施           14823 non-null object
洗手间数量          14783 non-null float64
床的数量           14796 non-null float64
床的类型           14823 non-null int64
卧室数量           14810 non-null float64
取消条款           14823 non-null int64
所在城市           14823 non-null int64
清洁费            14823 non-null int64
首次评论日期         11661 non-null object
房主是否有个人资料图片    14795 non-null object
房主身份是否验证       14795 non-null object
房主回复率          11165 non-null object
何时成为房主         14795 non-null object
是否支持随即预订       14823 non-null int64
最近评论日期         11668 non-null object
维度             14823 non-null float64
经度             14823 non-null float64
民宿周边           13460 non-null object
评论个数           14823 non-null int64
房产类型           14823 non-null int64
民宿评分           11496 non-null float64
房型    

In [6]:
test.head()

Unnamed: 0    数据ID  容纳人数  \
0           0  test_0     2   
1           1  test_1     2   
2           2  test_2     5   
3           3  test_3     6   
4           4  test_4     2   

                                                便利设施  洗手间数量  床的数量  床的类型  卧室数量  \
0  {TV,Internet,"Wireless Internet","Air conditio...    1.5   1.0     4   1.0   
1  {TV,Internet,"Wireless Internet","Air conditio...    2.0   1.0     4   1.0   
2  {TV,"Cable TV","Wireless Internet","Air condit...    1.0   3.0     4   2.0   
3  {"Cable TV",Internet,"Wireless Internet","Air ...    1.0   3.0     4   1.0   
4  {Internet,"Wireless Internet","Air conditionin...    1.0   1.0     4   1.0   

   取消条款  所在城市  清洁费      首次评论日期 房主是否有个人资料图片 房主身份是否验证 房主回复率      何时成为房主  \
0     2     1    1  2015-05-25           t        t  100%  2015-05-20   
1     2     3    1  2015-11-09           t        t  100%  2015-09-08   
2     1     4    1  2017-05-15           t        t  100%  2017-05-06   
3     2     4    1  2012-11-12           t        t   70%  2009-02-06   
4     0     3    1  2017-02-17           t        t  100%  2015-10-20   

   是否支持随即预订      最近评论日期         维度          经度               民宿周边  评论个数  房产类型  \
0         1  2017-01-01  41.849684  -87.676270             Pilsen    17    17   
1         0  2015-11-15  34.068613 -118.246455          Echo Park     2     0   
2         1  2017-09-25  40.701958  -73.917352           Bushwick    25     0   
3         0  2017-07-29  40.742959  -73.990820  Flatiron District    12     0   
4         0  2017-03-25  34.046473 -117.734095                NaN     2    17   

    民宿评分  房型  
0   97.0   1  
1  100.0   0  
2   88.0   0  
3   82.0   0  
4  100.0   1

## 3.2 描述性统计

In [7]:
train.describe()

Unnamed: 0          容纳人数         洗手间数量          床的数量          床的类型  \
count  59288.000000  59288.000000  59128.000000  59184.000000  59288.000000   
mean   29643.500000      3.152054      1.236183      1.712862      3.934068   
std    17115.115717      2.153512      0.582392      1.259406      0.427149   
min        0.000000      1.000000      0.000000      0.000000      0.000000   
25%    14821.750000      2.000000      1.000000      1.000000      4.000000   
50%    29643.500000      2.000000      1.000000      1.000000      4.000000   
75%    44465.250000      4.000000      1.000000      2.000000      4.000000   
max    59287.000000     16.000000      8.000000     18.000000      4.000000   

               卧室数量          取消条款          所在城市           清洁费      是否支持随即预订  \
count  59210.000000  59288.000000  59288.000000  59288.000000  59288.000000   
mean       1.265817      1.136216      3.291509      0.733791      0.262347   
std        0.849750      0.855095      1.181140      0.441979      0.439914   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000      0.000000      3.000000      0.000000      0.000000   
50%        1.000000      1.000000      4.000000      1.000000      0.000000   
75%        1.000000      2.000000      4.000000      1.000000      1.000000   
max       10.000000      4.000000      5.000000      1.000000      1.000000   

                 维度            经度          评论个数          房产类型          民宿评分  \
count  59288.000000  59288.000000  59288.000000  59288.000000  45893.000000   
mean      38.450807    -92.349408     20.935805      5.750101     94.076526   
std        3.080455     21.688276     37.950408      8.610262      7.807910   
min       33.338905   -122.511500      0.000000      0.000000     20.000000   
25%       34.128782   -118.341975      1.000000      0.000000     92.000000   
50%       40.663197    -76.995645      6.000000      0.000000     96.000000   
75%       40.746167    -73.954621     23.000000     17.000000    100.000000   
max       42.390437    -70.999166    605.000000     34.000000    100.000000   

                 房型            价格  
count  59288.000000  59288.000000  
mean       0.473030     68.968581  
std        0.554711     10.352019  
min        0.000000      0.000000  
25%        0.000000     62.094534  
50%        0.000000     67.813597  
75%        1.000000     75.313815  
max        2.000000    109.650628

In [8]:
test.describe()

Unnamed: 0          容纳人数         洗手间数量          床的数量          床的类型  \
count  14823.000000  14823.000000  14783.000000  14796.000000  14823.000000   
mean    7411.000000      3.167510      1.231584      1.702893      3.939688   
std     4279.175855      2.153925      0.580656      1.232870      0.411683   
min        0.000000      1.000000      0.000000      0.000000      0.000000   
25%     3705.500000      2.000000      1.000000      1.000000      4.000000   
50%     7411.000000      2.000000      1.000000      1.000000      4.000000   
75%    11116.500000      4.000000      1.000000      2.000000      4.000000   
max    14822.000000     16.000000      8.000000     16.000000      4.000000   

               卧室数量          取消条款          所在城市           清洁费      是否支持随即预订  \
count  14810.000000  14823.000000  14823.000000  14823.000000  14823.000000   
mean       1.265699      1.136814      3.297106      0.735209      0.262902   
std        0.861676      0.853313      1.184719      0.441237      0.440225   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000      0.000000      3.000000      0.000000      0.000000   
50%        1.000000      1.000000      4.000000      1.000000      0.000000   
75%        1.000000      2.000000      4.000000      1.000000      1.000000   
max       10.000000      4.000000      5.000000      1.000000      1.000000   

                 维度            经度         评论个数          房产类型          民宿评分  \
count  14823.000000  14823.000000  14823.00000  14823.000000  11496.000000   
mean      38.426567    -92.589982     20.75963      5.829454     94.030793   
std        3.079042     21.773036     37.33857      8.645638      7.950122   
min       33.343624   -122.510940      0.00000      0.000000     20.000000   
25%       34.123770   -118.343958      1.00000      0.000000     92.000000   
50%       40.656712    -77.001034      6.00000      0.000000     96.000000   
75%       40.745857    -73.954731     24.00000     17.000000    100.000000   
max       42.389907    -70.985047    449.00000     34.000000    100.000000   

                 房型  
count  14823.000000  
mean       0.466775  
std        0.554139  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        2.000000

## 3.3 重复值

In [9]:
train.duplicated().sum()

0

In [10]:
test.duplicated().sum()

0

## 3.4 缺失值

In [11]:
print(train.isnull().sum())

Unnamed: 0         0
数据ID               0
容纳人数               0
便利设施               0
洗手间数量            160
床的数量             104
床的类型               0
卧室数量              78
取消条款               0
所在城市               0
清洁费                0
首次评论日期         12702
房主是否有个人资料图片      160
房主身份是否验证         160
房主回复率          14641
何时成为房主           160
是否支持随即预订           0
最近评论日期         12672
维度                 0
经度                 0
民宿周边            5509
评论个数               0
房产类型               0
民宿评分           13395
房型                 0
价格                 0
dtype: int64


In [12]:
print(test.isnull().sum())

Unnamed: 0        0
数据ID              0
容纳人数              0
便利设施              0
洗手间数量            40
床的数量             27
床的类型              0
卧室数量             13
取消条款              0
所在城市              0
清洁费               0
首次评论日期         3162
房主是否有个人资料图片      28
房主身份是否验证         28
房主回复率          3658
何时成为房主           28
是否支持随即预订          0
最近评论日期         3155
维度                0
经度                0
民宿周边           1363
评论个数              0
房产类型              0
民宿评分           3327
房型                0
dtype: int64


# 4. 数据特征处理

## 4.1 便利设施  
对数据集中的“便利设施”这一列进行处理，变成新的一列“便利设施_num”——里面的信息为原列里面设施数量

In [13]:
train['便利设施_num'] = train['便利设施'].str.split(',')
int_equipments = []
for item in train['便利设施_num']:
    num = len(item)
    int_equipments.append(num)
train['便利设施_num'] = int_equipments


test['便利设施_num'] = test['便利设施'].str.split(',')
int_equipments_test = []
for item in test['便利设施_num']:
    num = len(item)
    int_equipments_test.append(num)
test['便利设施_num'] = int_equipments_test

为特定关键词创建新的二元列，并将其转换成字符串格式，以便后续使用LabelEncoder进行统一处理

In [14]:
for key_words in ['Internet', 'Kitchen', 'Pets', 'Air', 'Washer', 'Aid', 'Heating', 'Dryer', 'Essentials', 'Hangers', 'TV', 'parking', '24-hour']:
    train[key_words] = train['便利设施'].str.contains(key_words).astype('str')
    test[key_words] = test['便利设施'].str.contains(key_words).astype('str')

## 4.2 缺失值处理  

### 4.2.1民宿周边，使用-999进行缺失值填充，之后缺失值记为0，非缺失值记为1

In [15]:
train['民宿周边'] = train['民宿周边'].fillna(-999)
test['民宿周边'] = test['民宿周边'].fillna(-999)
surround = []
for value in train['民宿周边']:
    if value == -999:
        new_value = 0
    else:
        new_value = 1
    surround.append(new_value)
train['民宿周边'] = surround

surround_test = []
for value in test['民宿周边']:
    if value == -999:
        new_value = 0
    else:
        new_value = 1
    surround_test.append(new_value)
test['民宿周边'] = surround_test

针对某一些列，使用 f 来进行缺失值填充  
对另外的某一些列，需要进行Label编码，将字符串值转换成整数，以便用于机器模型的训练，对于每列，使用fit_transform方法在训练集上拟合编码器并对训练集和测试集进行编码

In [16]:
train['房主是否有个人资料图片'] = train['房主是否有个人资料图片'].fillna('f')
test['房主是否有个人资料图片'] = test['房主是否有个人资料图片'].fillna('f')

surround1 = []
for value in train['房主是否有个人资料图片']:
    if value == 'f':
        new_value = 0
    else:
        new_value = 1
    surround1.append(new_value)
train['房主是否有个人资料图片'] = surround1

surround1_test = []
for value in test['房主是否有个人资料图片']:
    if value == 'f':
        new_value = 0
    else:
        new_value = 1
    surround1_test.append(new_value)
test['房主是否有个人资料图片'] = surround1_test

In [17]:
train['房主身份是否验证'] = train['房主身份是否验证'].fillna('f')
test['房主身份是否验证'] = test['房主身份是否验证'].fillna('f')

surround2 = []
for value in train['房主身份是否验证']:
    if value == 'f':
        new_value = 0
    else:
        new_value = 1
    surround2.append(new_value)
train['房主身份是否验证'] = surround2

surround2_test = []
for value in test['房主身份是否验证']:
    if value == 'f':
        new_value = 0
    else:
        new_value = 1
    surround2_test.append(new_value)
test['房主身份是否验证'] = surround2_test

In [18]:
str_cols = ['Internet', 'Kitchen', 'Pets', 'Air', 'Washer', 'Aid', 'Heating', 'Dryer', 'Essentials', 'Hangers', 'TV', 'parking', '24-hour']

le = LabelEncoder()
for col in str_cols:
    train[col] = le.fit_transform(train[col])
    test[col] = le.fit_transform(test[col])

### 4.2.2 房主回复率  
将缺失值填充为20%（随意取的），再去掉最后的%字符，这样就将百分比值转换成浮点数值。

In [19]:
train['房主回复率'] = train['房主回复率'].fillna('20%')
test['房主回复率'] = test['房主回复率'].fillna('20%')
train['房主回复率'] = train['房主回复率'].str[:-1]
test['房主回复率'] = test['房主回复率'].str[:-1]

### 4.2.3 日期数据  
使用2019-1-1填充缺失值（随意选的），将日期时间列转换成日期时间数据类型，以便后续可以在日期时间特征中使用

In [20]:
train['首次评论日期'] = train['首次评论日期'].fillna('2019-1-1')
test['首次评论日期'] = test['首次评论日期'].fillna('2019-1-1')
train['最近评论日期'] = train['最近评论日期'].fillna('2019-1-1')
test['最近评论日期'] = test['最近评论日期'].fillna('2019-1-1')

time_cols = ['首次评论日期', '最近评论日期', '何时成为房主']
for col in time_cols:
    train[col] = pd.to_datetime(train[col])
    test[col] = pd.to_datetime(test[col])

### 4.2.4 将数据集划分成连续变量和分类变量，分别进行不同的处理  
将分类变量数据类型转换成category，以便后续处理  
连续变量：中位数  
分类变量：众数  
最后重置索引，以确保数据集中不再存在缺失值

In [21]:
cols = train.columns
cont_cols = ['容纳人数', '便利设施_num', '床的数量', '卧室数量', '洗手间数量', '维度', '经度', '评论个数', '民宿评分', '房主回复率']
cate_cols = ['Internet', 'Kitchen', 'Air', 'Pets', 'Washer', 'Aid', 'Heating', 'Dryer', 'Essentials', 'Hangers',
        'TV', 'parking', '24-hour', '床的类型', '民宿周边', '取消条款', '所在城市', '清洁费', '是否支持随即预订', 
        '房产类型', '房型', '房主是否有个人资料图片', '房主身份是否验证']
features = cont_cols + cate_cols
train[cate_cols] = train[cate_cols].astype('category')
test[cate_cols] = test[cate_cols].astype('category')

for col in cate_cols:
    train[col] = train[col].fillna(train[col].mode())
    test[col] = test[col].fillna(test[col].mode())

for col in ['洗手间数量', '床的数量', '卧室数量', '民宿评分']:
    train[col] = train[col].fillna(train[col].median())
    test[col] = test[col].fillna(test[col].median())

train = train.dropna().reset_index()

### 4.2.5 标准化处理（对连续变量进行特征缩放）  
标准差标准化：均值为0，标准差为1的分布  
四分位数范围标准化：对数据中的离群值不敏感  
Min-Max标准化：通常缩放到0-1之间  
使用fit方法对连续变量进行拟合，之后使用transform()方法进行特征缩放

In [22]:
X = train[features]

scaler = RobustScaler()
# scaler = StandardScaler()
# scaler = MinMaxScaler()

scaler.fit(X[cont_cols])

X[cont_cols] = scaler.transform(X[cont_cols])

X_test = test[features]
X_test[cont_cols] = scaler.transform(X_test[cont_cols])

y = train['价格'].astype('int')

In [23]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14823 entries, 0 to 14822
Data columns (total 33 columns):
容纳人数           14823 non-null float64
便利设施_num       14823 non-null float64
床的数量           14823 non-null float64
卧室数量           14823 non-null float64
洗手间数量          14823 non-null float64
维度             14823 non-null float64
经度             14823 non-null float64
评论个数           14823 non-null float64
民宿评分           14823 non-null float64
房主回复率          14823 non-null float64
Internet       14823 non-null category
Kitchen        14823 non-null category
Air            14823 non-null category
Pets           14823 non-null category
Washer         14823 non-null category
Aid            14823 non-null category
Heating        14823 non-null category
Dryer          14823 non-null category
Essentials     14823 non-null category
Hangers        14823 non-null category
TV             14823 non-null category
parking        14823 non-null category
24-hour        14823 non-null category
床的类型

In [25]:
import seaborn as sns
correlation_matrix = train.corr()
sns.heatmap(correlation_matrix, annot=True, cmap = 'coolwarm', linewidth=0.5)

# 5. 模型训练  

In [26]:
from sklearn.model_selection import train_test_split

X_train, X_valid, y_train, y_valid = train_test_split(X, y, random_state=2024, test_size=0.2)

In [27]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 47302 entries, 45953 to 7816
Data columns (total 33 columns):
容纳人数           47302 non-null float64
便利设施_num       47302 non-null float64
床的数量           47302 non-null float64
卧室数量           47302 non-null float64
洗手间数量          47302 non-null float64
维度             47302 non-null float64
经度             47302 non-null float64
评论个数           47302 non-null float64
民宿评分           47302 non-null float64
房主回复率          47302 non-null float64
Internet       47302 non-null category
Kitchen        47302 non-null category
Air            47302 non-null category
Pets           47302 non-null category
Washer         47302 non-null category
Aid            47302 non-null category
Heating        47302 non-null category
Dryer          47302 non-null category
Essentials     47302 non-null category
Hangers        47302 non-null category
TV             47302 non-null category
parking        47302 non-null category
24-hour        47302 non-null category
床

In [28]:
train.head()

index  Unnamed: 0     数据ID  容纳人数  \
0      0           0  train_0     4   
1      1           1  train_1     2   
2      2           2  train_2     4   
3      3           3  train_3     2   
4      4           4  train_4     3   

                                                便利设施  洗手间数量  床的数量 床的类型  卧室数量  \
0  {TV,"Cable TV",Internet,"Wireless Internet","A...    1.5   3.0    4   2.0   
1  {TV,"Wireless Internet",Kitchen,"Free parking ...    1.0   1.0    4   1.0   
2  {TV,"Air conditioning",Kitchen,Heating,"Smoke ...    1.0   2.0    4   0.0   
3                                                 {}    1.0   1.0    4   1.0   
4  {Internet,"Wireless Internet","Air conditionin...    1.0   1.0    4   1.0   

  取消条款 所在城市 清洁费     首次评论日期 房主是否有个人资料图片 房主身份是否验证 房主回复率     何时成为房主 是否支持随即预订  \
0    2    4   1 2015-05-07           1        1    20 2015-02-25        0   
1    2    4   1 2016-07-02           1        1    20 2009-10-27        1   
2    2    4   1 2017-07-01           1        0   100 2017-06-29        1   
3    0    5   1 2019-01-01           1        1    20 2013-03-19        0   
4    1    4   1 2014-04-30           1        1   100 2011-07-30        0   

      最近评论日期         维度          经度 民宿周边  评论个数 房产类型  民宿评分 房型         价格  \
0 2016-06-26  34.109039 -118.273390    1    12    0  97.0  0  64.918531   
1 2016-07-31  40.812897  -73.919163    1     6    0  87.0  0  54.918531   
2 2017-07-31  40.737643  -73.953309    1     4    0  80.0  0  73.219281   
3 2019-01-01  37.759935 -122.420558    1     0    0  96.0  1  64.093909   
4 2016-05-22  40.683363  -73.949490    1    16    0  99.0  0  68.454901   

   便利设施_num Internet Kitchen Pets Air Washer Aid Heating Dryer Essentials  \
0        17        1       1    0   1      1   0       1     1          1   
1        18        1       1    1   0      1   0       1     1          1   
2        11        0       1    0   1      0   0       1     0          1   
3         1        0       0    0   0      0   0       0     0          0   
4        13        1       1    1   1      0   0       1     0          1   

  Hangers TV parking 24-hour  
0       1  1       0       0  
1       1  1       1       0  
2       0  1       0       0  
3       0  0       0       0  
4       0  0       1       0

## RF模型

In [29]:
from sklearn.ensemble import RandomForestRegressor 
# 创建随机森林回归模型  
rf = RandomForestRegressor(n_estimators=100, random_state=42)  
rf.fit(X_train, y_train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=False, random_state=42, verbose=0,
                      warm_start=False)

In [30]:
y_pred_valid = rf.predict(X_valid)
y_pred_valid

array([65.34, 92.46, 69.65, ..., 65.74, 73.98, 80.03])

In [31]:
result_df = X_valid.copy()
result_df['价格'] = y_valid
result_df['预测价格'] = y_pred_valid
result_df.head()

容纳人数  便利设施_num  床的数量  卧室数量  洗手间数量        维度        经度      评论个数  \
31915   0.0 -0.222222   0.0  -1.0    0.0 -0.992243 -0.931753 -0.045455   
53380   2.0  0.222222   2.0   2.0    2.0 -1.001483 -0.935618 -0.272727   
12517   0.5 -0.111111   0.0   0.0    0.0  0.019307  0.068810 -0.227273   
10408   0.0  1.000000   0.0   0.0    1.0  0.006185  0.069217 -0.181818   
52012   0.0 -0.333333   0.0   0.0    0.0  0.008579  0.067725  0.227273   

           民宿评分  房主回复率 Internet Kitchen Air Pets Washer Aid Heating Dryer  \
31915 -0.666667    0.0        1       1   1    0      1   0       1     1   
53380  0.000000    0.0        1       1   1    1      1   0       1     1   
12517  0.666667   -1.0        1       1   1    0      1   0       1     1   
10408 -2.666667    0.0        1       1   1    0      1   0       1     1   
52012  0.333333   -1.0        1       1   1    0      0   0       1     0   

      Essentials Hangers TV parking 24-hour 床的类型 民宿周边 取消条款 所在城市 清洁费 是否支持随即预订  \
31915          0       1  1       0       0    4    0    2    3   1        0   
53380          1       0  1       1       0    4    1    2    3   1        0   
12517          1       1  1       0       0    4    1    0    4   1        0   
10408          1       0  1       0       0    4    1    1    4   1        0   
52012          1       1  1       0       1    4    1    2    4   0        0   

      房产类型 房型 房主是否有个人资料图片 房主身份是否验证  价格   预测价格  
31915    0  0           1        1  63  65.34  
53380   17  0           1        0  94  92.46  
12517    0  0           1        1  64  69.65  
10408    0  1           1        1  62  58.79  
52012    0  0           1        0  69  71.20

## CatBoost回归模型

In [32]:
# from catboost import CatBoostRegressor,Pool
# train_pool = Pool(X_train, y_train, cat_features = cate_cols)
# eval_pool = Pool(X_valid, y_valid, cat_features = cate_cols)

# model = CatBoostRegressor()
# model.fit(train_pool, eval_set = eval_pool)


![Image Name](https://cdn.kesci.com/upload/sc66weptdf.png?imageView2/0/w/1280/h/1280)  

![Image Name](https://cdn.kesci.com/upload/sc66y56h46.png?imageView2/0/w/1280/h/1280)  


预测测试集的价格

In [33]:
# y_pred_valid = model.predict(X_valid)
# y_pred_valid

创建一个新的DataFrame，将实际值、预测值和特征合并在一起

In [34]:
# result_df = X_valid.copy()
# result_df['价格'] = y_valid
# result_df['预测价格'] = y_pred_valid
# result_df.head()


![Image Name](https://cdn.kesci.com/upload/sc6709adq.png?imageView2/0/w/1280/h/1280)  


# 6. 模型评价

## RF模型

In [35]:
from sklearn.metrics import mean_squared_error, r2_score

mse = mean_squared_error(y_valid, y_pred_valid)
r2 = r2_score(y_valid, y_pred_valid)
print("mse的值为：", mse)
print("r2的值为：", r2)

mse的值为： 31.601622044647385
r2的值为： 0.70765507451574


In [36]:
predictions = rf.predict(X_test)
predictions

array([79.63, 69.82, 67.83, ..., 68.18, 76.54, 59.6 ])

In [37]:
new_result_df = X_test.copy()
new_result_df['预测价格'] = predictions
new_result_df.head()

容纳人数  便利设施_num  床的数量  卧室数量  洗手间数量        维度        经度      评论个数      民宿评分  \
0   0.0  0.222222   0.0   0.0    0.5  0.179369 -0.240588  0.500000  0.166667   
1   0.0 -0.555556   0.0   0.0    1.0 -0.996344 -0.929297 -0.181818  0.666667   
2   1.5  0.888889   2.0   1.0    0.0  0.005949  0.069383  0.863636 -1.333333   
3   2.0 -0.888889   2.0   0.0    0.0  0.012144  0.067728  0.272727 -2.333333   
4   0.0  0.000000   0.0   0.0    0.0 -0.999690 -0.917754 -0.181818  0.666667   

   房主回复率 Internet Kitchen Air Pets Washer Aid Heating Dryer Essentials  \
0  0.000        1       1   1    0      1   0       1     1          1   
1  0.000        1       1   1    0      1   0       1     1          1   
2  0.000        1       1   1    1      0   0       1     0          1   
3 -0.375        1       1   1    0      0   0       1     0          0   
4  0.000        1       0   1    0      0   0       1     0          1   

  Hangers TV parking 24-hour 床的类型 民宿周边 取消条款 所在城市 清洁费 是否支持随即预订 房产类型 房型  \
0       1  1       0       0    4    1    2    4   1        0    0  0   
1       0  1       0       0    4    1    2    3   1        0    0  0   
2       1  1       0       0    4    1    1    4   1        1    0  0   
3       0  1       0       0    4    1    2    4   1        0    0  0   
4       1  0       1       0    4    0    0    3   1        0   17  1   

  房主是否有个人资料图片 房主身份是否验证   预测价格  
0           1        1  79.63  
1           1        1  69.82  
2           1        1  67.83  
3           1        1  74.61  
4           1        1  57.14

## catboost模型

In [38]:
# from sklearn.metrics import mean_squared_error, r2_score

# mse = mean_squared_error(y_valid, y_pred_valid)
# r2 = r2_score(y_valid, y_pred_valid)
# print("mse的值为：", mse)
# print("r2的值为：", r2)


![Image Name](https://cdn.kesci.com/upload/sc6721zd8s.png?imageView2/0/w/1280/h/1280)  


In [39]:
# predictions = model.predict(X_test)

创建一个新的DataFrame，将实际值、预测值和特征合并在一起

In [40]:
# new_result_df = X_test.copy()
# new_result_df['预测价格'] = predictions
# new_result_df.head()


![Image Name](https://cdn.kesci.com/upload/sc673g69bm.png?imageView2/0/w/1280/h/1280)  


# 结论  
RF模型：  
		mse的值为： 31.601622044647385  
		r2的值为： 0.70765507451574  
CatBoost模型：  
		mse的值为： 29.79467079557383  
		r2的值为： 0.7243710844571845